File: algorithms/304.Range_Sum_Query_2D_-_Immutable/Range_Sum_Query_2D_-_Immutable.java
public class NumMatrix {

private int[][] sum2Origin = null;
public NumMatrix(int[][] matrix) {
    if (matrix == null || matrix.length <= 0)
        return;
    sum2Origin = new int[matrix.length + 1][matrix[0].length + 1];
    for (int i = 1; i <= matrix.length; i ++)
        for (int j = 1; j <= matrix[0].length; j ++)
            sum2Origin[i][j] = sum2Origin[i][j - 1] + sum2Origin[i - 1][j] - sum2Origin[i - 1][j - 1] + matrix[i - 1][j - 1];
    return;
}

public int sumRegion(int row1, int col1, int row2, int col2) {
    if (sum2Origin == null)
        return 0;
    return sum2Origin[row2 + 1][col2 + 1] - sum2Origin[row2 + 1][col1] - sum2Origin[row1][col2 + 1] + sum2Origin[row1][col1];

}

}

File: algorithms/304.Range_Sum_Query_2D_-_Immutable/Range_Sum_Query_2D_-_Immutable.py
<<<<<<< HEAD
class NumMatrix(object):
      def __init__(self, matrix):
          if matrix is None or not matrix:
              return
          n, m = len(matrix), len(matrix[0])
          self.sums = [ [0 for j in xrange(m+1)] for i in xrange(n+1) ]
          for i in xrange(1, n+1):
              for j in xrange(1, m+1):
                  self.sums[i][j] = matrix[i-1][j-1] + self.sums[i][j-1] + self.sums[i-1][j] - self.sums[i-1][j-1]


      def sumRegion(self, row1, col1, row2, col2):
          row1, col1, row2, col2 = row1+1, col1+1, row2+1, col2+1
          return self.sums[row2][col2] - self.sums[row2][col1-1] - self.sums[row1-1][col2] + self.sums[row1-1][col1-1]
=======
class NumMatrix(object):
      def __init__(self, matrix):
          if matrix is None or not matrix:
              return
          n, m = len(matrix), len(matrix[0])
          self.sums = [ [0 for j in xrange(m+1)] for i in xrange(n+1) ]
          for i in xrange(1, n+1):
              for j in xrange(1, m+1):
                  self.sums[i][j] = matrix[i-1][j-1] + self.sums[i][j-1] + self.sums[i-1][j] - self.sums[i-1][j-1]


      def sumRegion(self, row1, col1, row2, col2):
          row1, col1, row2, col2 = row1+1, col1+1, row2+1, col2+1
          return self.sums[row2][col2] - self.sums[row2][col1-1] - self.sums[row1-1][col2] + self.sums[row1-1][col1-1]
>>>>>>> 6200c8704614e918c8bfa5357c648dd1b4f7eb74
